### PR TITLE
persist secrets

### PR DESCRIPTION
**Pull request by mwilde345**
Ticket(s): FE-6132

## Problem
* No tests to validate the behavior of the authNZ middleware

## Solution

* add them. 
  * Includes a combination of stubbing out some class methods (e.g. `AccountKey.get`) and also stubbing out the 'edge' (`fetch` and `fs`)

## Result

* We know what functionality the authNZ middleware provides

## Testing
Screenshot 2024-11-20 at 1 17 16 PM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -66,7 +66,7 @@
     "pretest": "npm run format && npm run lint",
     "lint": "eslint . --fix",
     "test": "mocha --recursive ./test --require ./test/mocha-root-hooks.mjs --reporter spec --reporter mocha-junit-reporter",
-    "test:local": "mocha --recursive ./test --require ./test/mocha-root-hooks.mjs",
+    "test:local": "mocha --recursive ./test/authNZ.mjs --require ./test/mocha-root-hooks.mjs",
```
Comment (echo-bravo-yahoo) [inline package.json:69]: remove!

File: src/cli.mjs [hunk 2]
```diff
@@ -94,7 +94,12 @@ function buildYargs(argvInput) {
         alias: "d",
         type: "string",
         description: "a database path, including region",
-        // required: true,
+      },
+      role: {
+        alias: "r",
+        type: "string",
+        description: "a role",
+        default: "admin",
```
Comment (echo-bravo-yahoo) [inline src/cli.mjs:102]: does this apply for every CLI command, or should it be shared to a smaller set of commands?

File: test/helpers.mjs [hunk 3]
```diff
@@ -2,8 +2,8 @@ import { join } from "node:path";
 
 // small helper for sinon to wrap your return value
 // in the shape fetch would return it from the network
-export function f(returnValue) {
-  return { json: async () => returnValue };
+export function f(returnValue, status) {
+  return { json: async () => returnValue, status: status || 200 };
```
Comment (echo-bravo-yahoo) [inline test/helpers.mjs:6]: 👍 love it

File: test/authNZ.mjs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      }),
+    };
+  };
+
+  beforeEach(() => {
+    container = setupContainer();
+    container.register({
+      accountClient: awilix.asFunction(mockAccountClient).scoped(),
+      accountCreds: awilix.asClass(AccountKey).scoped(),
+      secretCreds: awilix.asClass(SecretKey).scoped(),
```
Comment (echo-bravo-yahoo) [inline test/authNZ.mjs:37]: if these are good defaults for all/new tests (and they look to me like they are?), you should move these registrations into `./config/setup-test-container.mjs`.

File: test/authNZ.mjs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    const exit = scope.resolve("exit");
+    const accountCreds = scope.resolve("accountCreds");
+
+    accountCreds.get = stub().throws(new InvalidCredsError());
+
+    await authNZMiddleware(argv);
+    expect(logger.stderr.args[0][0]).to.include("not signed in or has expired");
+    expect(logger.stdout.args[0][0]).to.include(
+      "To sign in, run:\n\nfauna login --profile test-profile\n",
+    );
```
Comment (echo-bravo-yahoo) [inline test/authNZ.mjs:63]: you can get this effect from this (more readable) syntax. sinon.match will do a substring search.

```
expect(logger.stderr).to.have.been.calledWith(sinon.match("not signed in or has expired"))
```

however, if you know what all of the output should look like, i added `scope.resolve("stdoutStream")` recently, which you can use like [this](https://github.com/fauna/fauna-shell/blob/v3/test/shell.mjs#L123) to make assertions against the whole output.

File: test/authNZ.mjs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  it("should refresh session if account key is invalid", async function () {
+    const argv = { authRequired: true, profile: "test-profile" };
+    const scope = container.createScope();
+
+    await run("db list", scope);
+    const accountCreds = scope.resolve("accountCreds");
+    accountCreds.save = stub();
+
+    const fs = scope.resolve("fs");
+    fs.readFileSync.withArgs(sinon.match(/secret_keys/)).returns("{}");
```
Comment (echo-bravo-yahoo) [inline test/authNZ.mjs:76]: +1, love the usage of withArgs and the matcher.

File: test/authNZ.mjs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+      .withArgs(sinon.match(/access_keys/))
+      .returns(validAccessKeyFile);
+
+    const accountClient = scope.resolve("accountClient");
+    accountClient.whoAmI.onFirstCall().throws(new InvalidCredsError());
+
+    await authNZMiddleware(argv);
+    expect(accountClient.refreshSession.calledOnce).to.be.true;
+    expect(accountCreds.save.calledOnce).to.be.true;
+    expect(accountCreds.save.args[0][0].creds).to.deep.equal({
```
Comment (echo-bravo-yahoo) [inline test/authNZ.mjs:87]: avoid the `args[0][0]` pattern when it's otherwise expressable.

**Review by echo-bravo-yahoo — CHANGES_REQUESTED** [review]

File: src/cli.mjs (hunk 2 above)
Comment (mwilde345) [reply src/cli.mjs:104]: everything that will be using the middleware, i.e. everything that needs a db key.

**Review by echo-bravo-yahoo — APPROVED** [review]

**Pull request merged**